chanfitness/chanevent.go

```diff
@@ -0,0 +1,150 @@
+package chanfitness
+
+import (
+	"time"
+)
+
+type eventType int
+
+const (
+	peerOnlineEvent eventType = iota
+	peerOfflineEvent
+)
+
+// String provides string representations of channel events.
+func (e eventType) String() string {
+	switch e {
+	case peerOnlineEvent:
+		return "peer_online"
+
+	case peerOfflineEvent:
+		return "peer_offline"
+	}
+
+	return "unknown"
+}
+
+// channelEvent is a a timestamped event which is observed on a per channel
+// basis.
+type channelEvent struct {
+	timestamp time.Time
+	eventType eventType
+}
+
+// chanEventLog stores all events that have occurred over a channel's lifetime.
+type chanEventLog struct {
+	// id is the uint64 of the short channel ID.
+	id uint64
+
+	// peer is the compressed public key of the peer being monitored.
+	peer [33]byte
+
+	// events is a log of timestamped events observed for the channel.
+	events []*channelEvent
+
+	// now is expected to return the current time. It is supplied as an
+	// external function to enable deterministic unit tests.
+	now func() time.Time
+
+	// openedAt tracks the first time this channel was seen, this is not
+	// necessarily the time that it confirmed on chain because channel events
+	// are not persisted at present.
+	openedAt time.Time
+
+	// closedAt is the time that the channel was closed.
+	closedAt time.Time
+}
+
+// add appends an event with the given type and current time to the event log.
+// If the channel is already closed, it returns early without adding an event.
+func (e *chanEventLog) add(eventType eventType) {
+	if !e.closedAt.IsZero() {
```

Comment:
can add the `// If the channel is already closed, it returns early without adding an event.` also here